In [1]:
# 라이브러리 불러오기
import pandas as pd
# 경고창 무시 
import warnings
warnings.filterwarnings("ignore")

# 1. 상장폐지기업정리

In [2]:
# 2021년까지 코스닥에 상장된 모든 기업 불러오기
상장폐지_df = pd.read_csv('./datasets/ts2000/코스닥2021년까지상페기업.csv', encoding='cp949')
상장폐지_df

회사명  거래소코드     회계년도 상장폐지일  소속코드  업종코드
0      (주)CMG제약  58820  2011/12   NaN     5     1
1      (주)CMG제약  58820  2012/12   NaN     5     1
2      (주)CMG제약  58820  2013/12   NaN     5     1
3      (주)CMG제약  58820  2014/12   NaN     5     1
4      (주)CMG제약  58820  2015/12   NaN     5     1
...         ...    ...      ...   ...   ...   ...
15897   흥구석유(주)  24060  2017/12   NaN     5     1
15898   흥구석유(주)  24060  2018/12   NaN     5     1
15899   흥구석유(주)  24060  2019/12   NaN     5     1
15900   흥구석유(주)  24060  2020/12   NaN     5     1
15901   흥구석유(주)  24060  2021/12   NaN     5     1

[15902 rows x 6 columns]

In [3]:
# Dart에서 직접 확인한 데이터를 메모장에 저장하여 불러오는 코드
상장폐지사유_df = pd.read_csv('./상장폐지사유.txt', sep=',' , engine='python')
상장폐지사유_df.reset_index(inplace=True)
상장폐지사유_df.columns = ['회사명', '상장폐지사유']
상장폐지사유_df

회사명                                       상장폐지사유
0       (주)다함이텍                                         자본잠식
1     (주)지아이바이오                                       감사의견거절
2         (주)금빛                                       해산사유발생
3     (주)나노트로닉스   기업의 계속성 및 경영의 투명성 등을 종합적으로 고려하여 상장폐지기준에 해당
4       (주)네오퍼플                                       감사의견거절
..          ...                                          ...
270  한화화인케미칼(주)                                          피합병
271   현대정보기술(주)                                          피합병
272    현대페인트(주)                                       감사의견거절
273   현대하이스코(주)                                          피합병
274     현진소재(주)                                       감사의견거절

[275 rows x 2 columns]

In [4]:
# "(주)"를 제거하는 함수 정의
def remove_joo(text):
    if '(주)' in text:
        return text.replace('(주)', '')
    else:
        return text

# '회사명' 컬럼에 있는 "(주)" 제거
상장폐지_df['회사명'] = 상장폐지_df['회사명'].apply(remove_joo)
상장폐지사유_df['회사명'] = 상장폐지사유_df['회사명'].apply(remove_joo)

In [5]:
# 코스닥 상장폐지된 기업 확인
상장폐지_df = 상장폐지_df[~상장폐지_df['상장폐지일'].isna()][['회사명','거래소코드','상장폐지일']].drop_duplicates()
상장폐지_df

회사명   거래소코드       상장폐지일
146       구스앤홈  329050  2021/04/13
231      극동자동화  272420  2018/11/15
246         금빛   45890  2021/12/20
256      금오하이텍  165270  2017/06/07
337     나노트로닉스   10670  2014/04/11
...        ...     ...         ...
15377  피아이첨단소재  178920  2021/08/09
15627     한솔신텍   99660  2018/07/09
15678   한일네트웍스   46110  2022/10/05
15798   현대정보기술   26180  2019/07/17
15817     현진소재   53660  2022/04/26

[289 rows x 3 columns]

In [6]:
# 상장폐지 데이터와 상장폐지사유 데이터를 회사명을 기준으로 merge
상장폐지사유merge_df = pd.merge(상장폐지_df, 상장폐지사유_df, how='left', on="회사명")
상장폐지사유merge_df

회사명   거래소코드       상장폐지일  \
0       구스앤홈  329050  2021/04/13   
1      극동자동화  272420  2018/11/15   
2         금빛   45890  2021/12/20   
3      금오하이텍  165270  2017/06/07   
4     나노트로닉스   10670  2014/04/11   
..       ...     ...         ...   
284  피아이첨단소재  178920  2021/08/09   
285     한솔신텍   99660  2018/07/09   
286   한일네트웍스   46110  2022/10/05   
287   현대정보기술   26180  2019/07/17   
288     현진소재   53660  2022/04/26   

                                                상장폐지사유  
0                                                  NaN  
1                                                  NaN  
2                                               해산사유발생  
3                                                  NaN  
4           기업의 계속성 및 경영의 투명성 등을 종합적으로 고려하여 상장폐지기준에 해당  
..                                                 ...  
284                                              코스피상장  
285   발행한 어음 또는 수표가 주거래은행에 의하여 최종부도로 결정되거나 거래은행에 의한...  
286                                             상장폐지신청  
287                                                피합병  
288                                             감사의견거절  

[289 rows x 4 columns]

In [7]:
# 공백제거
상장폐지사유merge_df['상장폐지사유'] = 상장폐지사유merge_df['상장폐지사유'].str.strip()

In [8]:
# 상장폐지사유 결측치 확인
상장폐지사유merge_df['상장폐지사유'].isna().sum()

83

In [9]:
# 결측치 83개 모두 코넥스 데이터
상장폐지사유merge_df['상장폐지사유'].fillna('코넥스', inplace=True)

In [10]:
# 상장폐지사유들 확인해보기
상장폐지사유merge_df['상장폐지사유'].value_counts()

코넥스                                                                                       83
감사의견거절                                                                                    76
피합병                                                                                       26
자본전액잠식                                                                                    14
기업의 계속성 및 경영의 투명성 등을 종합적으로 고려하여 상장폐지기준에 해당한다고 결정                                          14
코스피상장                                                                                     10
상장폐지신청                                                                                     7
기업의 계속성 및 경영의 투명성 등을 종합적으로 고려하여 상장폐지기준에 해당                                                 5
발행한 어음 또는 수표가 주거래은행에 의하여 최종부도로 결정되거나 거래은행에 의한 거래정지                                         4
횡령 배임                                                                                      4
기업의 계속성 경영의 투명성 또는 그 밖에 코스닥시장의 건전성 등을 종합적으로 고려하여 상장폐지기준에 해당한다고 결정     

In [11]:
for idx, row in 상장폐지사유merge_df.iterrows():
    if '감사' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '감사의견거절'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '자본' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '자본잠식'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '해산' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '파산'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '파산' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '파산'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '보고서' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '보고서미제출'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '피합' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '피합병'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '신청' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '자진상장폐지'
for idx, row in 상장폐지사유merge_df.iterrows():
    if '자진' in row['상장폐지사유']:
        상장폐지사유merge_df.at[idx, '상장폐지사유'] = '자진상장폐지'

In [12]:
# 상장폐지 정의에 따른 1, 0 분류 코드
def change(x):
    if x == '피합병' or x == '자진상장폐지' or x == '코스피상장' or x == '코넥스':
        return 0
    else:
        return 1

상장폐지사유merge_df['상장폐지'] = 상장폐지사유merge_df['상장폐지사유'].apply(change)

In [13]:
# 상장폐지 기업 확인
상장폐지사유merge_df['상장폐지'].value_counts()

1    159
0    130
Name: 상장폐지, dtype: int64

### 2022년도까지 상장폐지 기업수 : 289개
### 부도정의로 분류한 결과 1: 159개 0: 130개

# 2. 연결데이터, 개별데이터 결합

In [14]:
# 성장성
성장성_연결 = pd.read_csv("./datasets/ts2000/kosdaq_성장성_연결.csv", encoding='cp949')
# 수익성
수익성_연결 = pd.read_csv("./datasets/ts2000/kosdaq_수익성_연결.csv", encoding='cp949')
# 안정성
안정성_연결 = pd.read_csv("./datasets/ts2000/kosdaq_안정성_연결.csv", encoding='cp949')
# 활동성
활동성_연결 = pd.read_csv("./datasets/ts2000/kosdaq_활동성_연결.csv", encoding='cp949')

In [15]:
# 컬럼명 변경
수익성_연결.rename(columns={'회사명' : '회사명1', '거래소코드' : '거래소코드1', '회계년도' : '회계년도1'}, inplace=True)
안정성_연결.rename(columns={'회사명' : '회사명2', '거래소코드' : '거래소코드2', '회계년도' : '회계년도2'}, inplace=True)
활동성_연결.rename(columns={'회사명' : '회사명3', '거래소코드' : '거래소코드3', '회계년도' : '회계년도3'}, inplace=True)

In [16]:
# 성장, 수익, 안정, 활동 결합
kosdaq_연결 = pd.concat([성장성_연결, 수익성_연결, 안정성_연결, 활동성_연결], axis=1)
kosdaq_연결.drop(['회사명1', '거래소코드1', '회계년도1','회사명2', '거래소코드2', '회계년도2','회사명3', '거래소코드3', '회계년도3'], axis=1, inplace=True)

In [17]:
# 컬럼명 정리 ( 특수문자제거 )
kosdaq_연결.columns = kosdaq_연결.columns.str.replace('연결','')
kosdaq_연결.columns = kosdaq_연결.columns.str.replace('원','')
kosdaq_연결.columns = kosdaq_연결.columns.str.replace('IFRS','')
kosdaq_연결.columns = kosdaq_연결.columns.str.replace('(','')
kosdaq_연결.columns = kosdaq_연결.columns.str.replace(')','')
kosdaq_연결.columns = kosdaq_연결.columns.str.replace(' ','')

In [18]:
# 연결재무제표 결측치 확인
kosdaq_연결.isna().sum()

회사명                  0
거래소코드                0
회계년도                 0
총자본증가율            6025
유형자산증가율           6025
투자부동산증가율          6025
비유동자산증가율          6025
유동자산증가율           6025
재고자산증가율           6025
자기자본증가율           6025
매출액증가율            6025
순이익증가율            6025
정상영업이익증가율         6025
매출액총이익률           6025
매출액순이익률           6025
총자본순이익률           6025
자기자본순이익률          6025
매출가대매출액비율         6025
영업비용대영업수익비율       6025
광고선전비대매출액비율       6025
기업순이익률            6025
R&D투자효율           6025
금융비용대총비용비율        6025
이자부담률             6025
1주당매출액            6025
1주당순이익            6025
1주당CASHFLOW       6025
1주당정상영업이익         6025
유동자산구성비율          6025
재고자산대유동자산비율       6025
유동자산대비유동자산비율      6025
당좌자산구성비율          6025
비유동자산구성비율         6025
자기자본구성비율          6025
타인자본구성비율          6025
비유동비율             6025
유동비율              6025
당좌비율              6025
현금비율              6025
매입채무대재고자산비율       6025
부채비율              6025
유동부채비율            6025
단기차입금대총차입금비율      6025
순운전자본비율    

## 연결 결측치 6025개

In [19]:
# 성장성
성장성_개별 = pd.read_csv("./datasets/ts2000/kosdaq_성장성_개별.csv", encoding='cp949')
# 수익성
수익성_개별 = pd.read_csv("./datasets/ts2000/kosdaq_수익성_개별.csv", encoding='cp949')
# 안정성
안정성_개별 = pd.read_csv("./datasets/ts2000/kosdaq_안정성_개별.csv", encoding='cp949')
# 활동성
활동성_개별 = pd.read_csv("./datasets/ts2000/kosdaq_활동성_개별.csv", encoding='cp949')

In [20]:
# 컬럼명 변경
수익성_개별.rename(columns={'회사명' : '회사명1', '거래소코드' : '거래소코드1', '회계년도' : '회계년도1'}, inplace=True)
안정성_개별.rename(columns={'회사명' : '회사명2', '거래소코드' : '거래소코드2', '회계년도' : '회계년도2'}, inplace=True)
활동성_개별.rename(columns={'회사명' : '회사명3', '거래소코드' : '거래소코드3', '회계년도' : '회계년도3'}, inplace=True)

In [21]:
# 성장, 수익, 안정, 활동 결합
kosdaq_개별 = pd.concat([성장성_개별, 수익성_개별, 안정성_개별, 활동성_개별], axis=1)
kosdaq_개별.drop(['회사명1', '거래소코드1', '회계년도1','회사명2', '거래소코드2', '회계년도2','회사명3', '거래소코드3', '회계년도3'], axis=1, inplace=True)

In [22]:
# 컬럼명 정리 ( 특수문자제거 )
kosdaq_개별.columns = kosdaq_개별.columns.str.replace('연결','')
kosdaq_개별.columns = kosdaq_개별.columns.str.replace('원','')
kosdaq_개별.columns = kosdaq_개별.columns.str.replace('IFRS','')
kosdaq_개별.columns = kosdaq_개별.columns.str.replace('(','')
kosdaq_개별.columns = kosdaq_개별.columns.str.replace(')','')
kosdaq_개별.columns = kosdaq_개별.columns.str.replace(' ','')

In [23]:
kosdaq_개별.isna().sum()

회사명                  0
거래소코드                0
회계년도                 0
총자본증가율            2116
유형자산증가율           2116
투자부동산증가율          2116
비유동자산증가율          2116
유동자산증가율           2116
재고자산증가율           2116
자기자본증가율           2116
매출액증가율            2116
순이익증가율            2116
정상영업이익증가율         2116
매출액총이익률           2116
매출액순이익률           2116
총자본순이익률           2116
자기자본순이익률          2116
매출가대매출액비율         2116
영업비용대영업수익비율       2116
광고선전비대매출액비율       2116
기업순이익률            2116
R&D투자효율           2116
금융비용대총비용비율        2116
이자부담률             2116
1주당매출액            2116
1주당순이익            2116
1주당CASHFLOW       2116
1주당정상영업이익         2116
유동자산구성비율          2116
재고자산대유동자산비율       2116
유동자산대비유동자산비율      2116
당좌자산구성비율          2116
비유동자산구성비율         2116
자기자본구성비율          2116
타인자본구성비율          2116
비유동비율             2116
유동비율              2116
당좌비율              2116
현금비율              2116
매입채무대재고자산비율       2116
부채비율              2116
유동부채비율            2116
단기차입금대총차입금비율      2116
순운전자본비율    

## 개별 결측치 2116개

In [24]:
filled_kosdaq = kosdaq_연결.combine_first(kosdaq_개별)

In [25]:
filled_kosdaq

회사명  거래소코드     회계년도  총자본증가율  유형자산증가율  투자부동산증가율  비유동자산증가율  유동자산증가율  \
0      (주)CMG제약  58820  2011/12    1.98    -5.90      0.00    -12.89    17.64   
1      (주)CMG제약  58820  2012/12  109.15    -4.12      0.00    -16.43   207.08   
2      (주)CMG제약  58820  2013/12   -2.09    16.09      0.00     77.05   -18.89   
3      (주)CMG제약  58820  2014/12    3.86    74.30      0.00     68.15   -25.93   
4      (주)CMG제약  58820  2015/12   -2.64    -5.37      0.00      1.66    -7.17   
...         ...    ...      ...     ...      ...       ...       ...      ...   
17393   흥구석유(주)  24060  2018/12   -5.89    -0.36     -6.55     -2.09   -21.19   
17394   흥구석유(주)  24060  2019/12   11.24    -7.00      6.36     -2.63    80.74   
17395   흥구석유(주)  24060  2020/12    0.07   -13.92     -0.30     -8.21    22.41   
17396   흥구석유(주)  24060  2021/12   -1.28     0.63     -0.30      0.18    -4.24   
17397   흥구석유(주)  24060  2022/12   13.66    -1.71    117.99     51.90   -67.31   

       재고자산증가율  자기자본증가율  ...  총자본회전률  자기자본회전률  타인자본회전률   매입채무회전률  유동자산회전률  \
0        -1.43    -2.02  ...    0.66     0.97     2.01     11.95     1.25   
1         7.73    75.17  ...    0.39     0.66     0.93      7.83     0.52   
2        56.29    55.10  ...    0.36     0.51     1.26      6.20     0.48   
3         6.37     0.01  ...    0.40     0.46     2.85      5.27     0.68   
4        11.04    -4.48  ...    0.47     0.56     2.88      5.66     0.99   
...        ...      ...  ...     ...      ...      ...       ...      ...   
17393   -46.09     0.99  ...    2.14     2.37    22.28   5128.70    11.71   
17394    96.40     9.31  ...    1.90     2.04    26.38  23497.93     8.58   
17395   -18.61     3.38  ...    1.37     1.47    21.30   9836.30     4.57   
17396    16.13    -0.88  ...    1.59     1.67    33.59   9998.01     4.88   
17397    18.35     1.51  ...    1.66     1.85    16.67  11249.20     8.34   

       당좌자산회전률  재고자산회전률  매출채권회전률  유형자산회전율  순운전자본회전률  
0         1.55     6.42     2.45     1.77      0.86  
1         0.57     5.71     1.85     1.71      0.42  
2         0.52     5.39     2.07     2.03      0.42  
3         0.80     4.77     1.86     1.54      0.56  
4         1.22     5.22     2.10     1.48      0.75  
...        ...      ...      ...      ...       ...  
17393    13.76    78.35    31.01     4.40      8.08  
17394     9.81    68.41    31.51     4.14      6.91  
17395     5.10    43.39    28.18     3.52      4.00  
17396     5.39    51.51    41.38     4.37      4.44  
17397    10.06    48.72    39.88     4.87      6.86  

[17398 rows x 59 columns]

In [26]:
filled_kosdaq.isna().sum()

회사명                  0
거래소코드                0
회계년도                 0
총자본증가율            2002
유형자산증가율           2002
투자부동산증가율          2002
비유동자산증가율          2002
유동자산증가율           2002
재고자산증가율           2002
자기자본증가율           2002
매출액증가율            2002
순이익증가율            2002
정상영업이익증가율         2002
매출액총이익률           2007
매출액순이익률           2007
총자본순이익률           2007
자기자본순이익률          2007
매출가대매출액비율         2007
영업비용대영업수익비율       2007
광고선전비대매출액비율       2007
기업순이익률            2007
R&D투자효율           2007
금융비용대총비용비율        2007
이자부담률             2007
1주당매출액            2007
1주당순이익            2007
1주당CASHFLOW       2007
1주당정상영업이익         2007
유동자산구성비율          2008
재고자산대유동자산비율       2008
유동자산대비유동자산비율      2008
당좌자산구성비율          2008
비유동자산구성비율         2008
자기자본구성비율          2008
타인자본구성비율          2008
비유동비율             2008
유동비율              2008
당좌비율              2008
현금비율              2008
매입채무대재고자산비율       2008
부채비율              2008
유동부채비율            2008
단기차입금대총차입금비율      2008
순운전자본비율    

## 연결재무제표 결측치 개별재무제표 값으로 대체 후 남은 결측치 약 2000개

-------------------------------------------------------

## 3. 생애주기, 업력 파생변수 결합

- 도입기 : 영업활동 - / 투자활동 - / 재무활동 +
- 성장기 : 영업활동 + / 투자활동 - / 재무활동 +
- 성숙기 : 영업활동 + / 투자활동 - / 재무활동 -
- Shake-Out : 도입/성장/성숙/쇠퇴기에 해당하지 않는 시기
- 쇠퇴기 : 영업활동 - / 투자활동 + / 재무활동 +/-

In [27]:
kosdaq_life = pd.read_csv('./datasets/ts2000/new_lifetime_kosdaq.csv', index_col=0)
kosdaq_life = kosdaq_life.iloc[:,3:]
kosdaq_concat = pd.concat([filled_kosdaq, kosdaq_life], axis=1)
kosdaq_life['기업생애주기'].value_counts()

성장기          4550
성숙기          4500
Shake-Out    3762
도입기          2942
쇠퇴기          1644
Name: 기업생애주기, dtype: int64

In [28]:
kosdaq_history = pd.read_csv('./datasets/ts2000/kosdaq_설립연도상장연도.csv', index_col=0)
kosdaq_history= kosdaq_history.iloc[:,3:]
kosdaq_concat_2 = pd.concat([kosdaq_concat, kosdaq_history], axis=1)
kosdaq_history

통계청 한국표준산업분류 10차(대분류)  설립연도  상장연도  경영기간  상장기간
0                       제조업  2001  2001    10    10
1                       제조업  2001  2001    11    11
2                       제조업  2001  2001    12    12
3                       제조업  2001  2001    13    13
4                       제조업  2001  2001    14    14
...                     ...   ...   ...   ...   ...
17393              도매 및 소매업  1966  1994    52    24
17394              도매 및 소매업  1966  1994    53    25
17395              도매 및 소매업  1966  1994    54    26
17396              도매 및 소매업  1966  1994    55    27
17397              도매 및 소매업  1966  1994    56    28

[17398 rows x 5 columns]

In [29]:
상장기간 = pd.read_csv('./datasets/kosdaq_상장기간.csv',index_col=0)
kosdaq_concat_2['상장기간'] = 상장기간['상장기간']

## 4. 상장폐지기업 결합

In [30]:
상장폐지사유merge_df = 상장폐지사유merge_df[['거래소코드','상장폐지일', '상장폐지사유', '상장폐지']]
상장폐지사유merge_df['상장폐지년도'] = pd.to_datetime(상장폐지사유merge_df['상장폐지일'], format='%Y/%m/%d').dt.strftime('%Y')
상장폐지사유merge_df['상장폐지년도'] = 상장폐지사유merge_df['상장폐지년도'].astype(int)
상장폐지사유merge_df

거래소코드       상장폐지일                                             상장폐지사유  \
0    329050  2021/04/13                                                코넥스   
1    272420  2018/11/15                                                코넥스   
2     45890  2021/12/20                                                 파산   
3    165270  2017/06/07                                                코넥스   
4     10670  2014/04/11         기업의 계속성 및 경영의 투명성 등을 종합적으로 고려하여 상장폐지기준에 해당   
..      ...         ...                                                ...   
284  178920  2021/08/09                                              코스피상장   
285   99660  2018/07/09  발행한 어음 또는 수표가 주거래은행에 의하여 최종부도로 결정되거나 거래은행에 의한 ...   
286   46110  2022/10/05                                             자진상장폐지   
287   26180  2019/07/17                                                피합병   
288   53660  2022/04/26                                             감사의견거절   

     상장폐지  상장폐지년도  
0       0    2021  
1       0    2018  
2       1    2021  
3       0    2017  
4       1    2014  
..    ...     ...  
284     0    2021  
285     1    2018  
286     0    2022  
287     0    2019  
288     1    2022  

[289 rows x 5 columns]

In [31]:
상장폐지사유merge_df['상장폐지'].value_counts()

1    159
0    130
Name: 상장폐지, dtype: int64

In [32]:
# 코스닥 데이터와 상장폐지데이터 merge
kosdaq = pd.merge(kosdaq_concat_2, 상장폐지사유merge_df, how='left', on='거래소코드')

In [33]:
# 상장폐지 결측치 0으로 대체 (정상기업)
kosdaq['상장폐지'] = kosdaq['상장폐지'].fillna(0)
kosdaq['상장폐지'] = kosdaq['상장폐지'].astype(int)

In [34]:
# 라벨이 1인 데이터수 확인
kosdaq['상장폐지'].value_counts()

0    16582
1      816
Name: 상장폐지, dtype: int64

In [35]:
# 라벨이 1인 기업수 확인
len(kosdaq[kosdaq['상장폐지'] == 1]['회사명'].unique())

159

## 5. 외국기업(거래소코드 앞자리가 9인 기업), 금융 기업, 12월 결산일이 아닌 데이터, 거래소코드가 2개인 기업 모두 제거

5-1) 외국 기업 (거래소코드 앞자리가 9인 기업)

지우는 이유 : 재무기준이 다르기 때문

In [36]:
# 종목코드를 6자리로 통일. ex) 990 -> 000990
def padding(x):
    x = str(x)
    return x.rjust(6, "0")
# True, False 반전
def func_not(x):
    return not x

In [37]:
 # 거래소코드 padding
kosdaq["거래소코드"] = kosdaq["거래소코드"].map(padding)

In [38]:
# 외국계 기업 확인
kosdaq[kosdaq['거래소코드'].str[:1] == '9']['회사명'].unique()

array(['(주)네오이뮨텍', '(주)소마젠', '(주)엑세스바이오인코퍼레이션', '3노드디지탈그룹유한공사',
       '글로벌에스엠테크리미티드', '로스웰인터내셔널유한회사', '미투젠(유)', '성융광전투자유한공사',
       '애머릿지코퍼레이션', '에스앤씨엔진그룹리미티드', '오가닉티코스메틱스홀딩스컴퍼니리미티드', '완리인터내셔널홀딩스',
       '웨이포트유한공사', '윙입푸드홀딩스', '이스트아시아홀딩스인베스트먼트리미티드', '잉글우드랩(주)',
       '차이나그레이트스타인터내셔널리미티드', '차이나크리스탈신소재홀딩스', '차이나하오란리사이클링유한공사',
       '컬러레이홀딩스', '케이만금세기차륜집단유한공사', '코오롱티슈진(주)', '헝셩그룹유한회사'], dtype=object)

In [39]:
## 외국계 기업 제거
def func_isnt_fcompany(x):
    if x[:1] == '9':
        return False
    else:
        return True

print("제거할 외국계기업 데이터 수 : ", len(kosdaq[kosdaq["거래소코드"].map(func_isnt_fcompany).map(func_not)]))
kosdaq = kosdaq[kosdaq["거래소코드"].map(func_isnt_fcompany)]
kosdaq.shape # 외국계기업 데이터 159개 제거

제거할 외국계기업 데이터 수 :  159


(17239, 73)

5-2) 금융 기업 데이터 제거

지우는 이유 : 재무기준이 다르기 때문

In [40]:
# 금융 및 보험업 상장폐지 기업 확인
kosdaq[kosdaq['통계청 한국표준산업분류 10차(대분류)'].str.contains('금융 및 보험업')]['회사명'].unique()

array(['글로벌텍스프리(주)', '인카금융서비스(주)', '해성산업(주)'], dtype=object)

In [41]:
# 금융 및 보험업 데이터 제거
print("제거할 금융 및 보험업 데이터 수 : ", len(kosdaq[kosdaq['통계청 한국표준산업분류 10차(대분류)'].str.contains('금융 및 보험업')]))
kosdaq = kosdaq[~kosdaq['통계청 한국표준산업분류 10차(대분류)'].str.contains('금융 및 보험업')]
kosdaq.shape # 금융 및 보험업 데이터 31개 제거

제거할 금융 및 보험업 데이터 수 :  31


(17208, 73)

5-3) 결산일이 12월이 아닌 데이터 제거

지우는 이유 : 비교 기준이 맞지 않기 때문

In [42]:
# 회계년도컬럼에서 월만 추출
kosdaq['회계년도월'] = pd.to_datetime(kosdaq['회계년도'], format='%Y/%m').dt.strftime('%m')
kosdaq['회계년도월'] = kosdaq['회계년도월'].astype(int)

In [43]:
# 회계년도가 12월이 아닌 데이터 제거
print("제거할 회계년도 12월이 아닌 데이터 수 : ", len(kosdaq[kosdaq['회계년도월'] != 12]))
kosdaq = kosdaq[kosdaq['회계년도월'] == 12]
kosdaq.drop(columns='회계년도월',axis=1, inplace=True)
kosdaq.shape # 회계년도가 12월이 아닌 데이터 121개 제거

제거할 회계년도 12월이 아닌 데이터 수 :  121


(17087, 73)

5-3) 거래소코드가 2개인 기업 제거

지우는 이유 : 데이터 통일

In [44]:
codecheck = kosdaq.groupby(['회사명'])['거래소코드'].nunique().reset_index()
codecheck.rename(columns = {'거래소코드':'거래소코드_개수'}, inplace=True)
codecheck = codecheck[codecheck['거래소코드_개수']>1]
codecheck

회사명  거래소코드_개수
286          (주)본느         2
386       (주)세화피앤씨         2
508     (주)아이엘사이언스         2
533        (주)알로이스         2
575       (주)에스에이티         2
661       (주)엔케이맥스         2
750       (주)원바이오젠         2
758       (주)원익피앤이         2
861         (주)인산가         2
887         (주)자비스         2
894         (주)정다운         2
974      (주)카이노스메드         2
1116    (주)티에스트릴리온         2
1166   (주)포인트엔지니어링         2
1230      (주)한국비엔씨         2
1267         (주)휴럼         2
1306       나무기술(주)         2
1441      소프트캠프(주)         2
1460      씨아이에스(주)         2
1517  에스케이오션플랜트(주)         2
1552    엠에프엠코리아(주)         2
1590         원텍(주)         2
1645       줌인터넷(주)         2
1694      패션플랫폼(주)         2

In [45]:
## 회사수와 거래소코드수 차이가나는것 확인
print("회사수 : ", len(kosdaq["회사명"].unique()))
print("거래소코드수 : ", len(kosdaq["거래소코드"].unique()))

회사수 :  1751
거래소코드수 :  1775


In [46]:
## 거래소 코드 중복기업 전처리
list_overlap_code = list()
for i in list(kosdaq["회사명"].unique()):
    if len(kosdaq[kosdaq["회사명"]==i]["거래소코드"].unique()) > 1:
        list_overlap_code.append(i)
print("거래소코드가 2개인 기업 수:", len(list_overlap_code), "개")
# list_overlap_code

거래소코드가 2개인 기업 수: 24 개


In [47]:
# 중복 거래소코드 제거
def func_overlap_code(x):
    if x == (kosdaq[kosdaq["회사명"]==company_name]["거래소코드"].iloc[-1]) :
        return True
    else:
        return False

list_companyname = list(kosdaq["회사명"].unique())


dfs = list()


for i in list_companyname:
    company_name = i
    #각 기업별로 데이터프레임 생성
    df_concat = kosdaq[kosdaq["회사명"]==i]
    df_concat = df_concat[df_concat["거래소코드"].map(func_overlap_code)] # 거래소코드 바뀌기 전의 거래소코드의 데이터는 삭제
    dfs.append(df_concat)
kosdaq_1 = pd.concat(dfs)
kosdaq_1.reset_index(drop=True, inplace=True)

In [48]:
# 데이터 제거 전
print(kosdaq.shape)
# 데이터 제거 후 
print(kosdaq_1.shape) # 158개 데이터 제거

(17087, 73)
(16929, 73)


In [49]:
# 거래소코드가 2개인 기업 잘 제거됐는지 한번 더 확인
codecheck = kosdaq_1.groupby(['회사명'])['거래소코드'].nunique().reset_index()
codecheck.rename(columns = {'거래소코드':'거래소코드_개수'}, inplace=True)
codecheck = codecheck[codecheck['거래소코드_개수']>1]
codecheck

Empty DataFrame
Columns: [회사명, 거래소코드_개수]
Index: []

In [50]:
# 회사명이 2개인 거래소코드 존재 확인 
list_overlap_code = list()
for i in list(kosdaq_1["거래소코드"].unique()):
    if len(kosdaq_1[kosdaq_1["거래소코드"]==i]["회사명"].unique()) > 1:
        list_overlap_code.append(i)
list_overlap_code

[]

# 6. 성장성 데이터 계산한 값으로 대체

In [51]:
kosdaq_성장성 = pd.read_csv('./datasets/ts2000/kosdaq_성장율_계산_edit.csv', index_col=0)

In [52]:
kosdaq_1 = kosdaq_1.iloc[:,13:]
kosdaq_1

매출액총이익률  매출액순이익률  총자본순이익률  자기자본순이익률  매출가대매출액비율  영업비용대영업수익비율  \
0        41.83    -8.50    -5.58     -8.28      58.17        97.89   
1        28.34   -40.97   -15.78    -26.84      71.66       125.87   
2        38.70     1.45     0.52      0.74      61.30        96.12   
3        37.07     0.69     0.28      0.32      62.93        98.06   
4        36.26   -13.50    -6.37     -7.62      63.74       109.51   
...        ...      ...      ...       ...        ...          ...   
16924     5.00     1.43     3.06      3.38      95.00        99.34   
16925     5.14     5.50    10.44     11.25      94.86        99.67   
16926     5.23     4.17     5.73      6.12      94.77       100.43   
16927     4.90     1.06     1.68      1.77      95.10       100.02   
16928     5.52     1.81     3.01      3.34      94.48        99.07   

       광고선전비대매출액비율  기업순이익률  R&D투자효율  금융비용대총비용비율  ...     기업생애주기  \
0             0.26   -5.05      0.0        0.72  ...        도입기   
1             0.14  -14.80      0.0        1.78  ...        도입기   
2             0.26    1.12      0.0        1.59  ...        쇠퇴기   
3             0.21    0.02      0.0       -0.64  ...        성장기   
4             1.46   -6.42      0.0       -0.10  ...        성장기   
...            ...     ...      ...         ...  ...        ...   
16924         0.00    3.03      0.0       -0.01  ...  Shake-Out   
16925         0.00   10.20      0.0       -0.12  ...        성숙기   
16926         0.00    5.57      0.0       -0.11  ...  Shake-Out   
16927         0.00    1.67      0.0       -0.01  ...        쇠퇴기   
16928         0.00    3.05      0.0        0.02  ...        성장기   

       통계청 한국표준산업분류 10차(대분류)  설립연도  상장연도  경영기간  상장기간  상장폐지일  상장폐지사유  상장폐지  \
0                        제조업  2001  2001    10    11    NaN     NaN     0   
1                        제조업  2001  2001    11    12    NaN     NaN     0   
2                        제조업  2001  2001    12    13    NaN     NaN     0   
3                        제조업  2001  2001    13    14    NaN     NaN     0   
4                        제조업  2001  2001    14    15    NaN     NaN     0   
...                      ...   ...   ...   ...   ...    ...     ...   ...   
16924               도매 및 소매업  1966  1994    52    25    NaN     NaN     0   
16925               도매 및 소매업  1966  1994    53    26    NaN     NaN     0   
16926               도매 및 소매업  1966  1994    54    27    NaN     NaN     0   
16927               도매 및 소매업  1966  1994    55    28    NaN     NaN     0   
16928               도매 및 소매업  1966  1994    56    29    NaN     NaN     0   

       상장폐지년도  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
...       ...  
16924     NaN  
16925     NaN  
16926     NaN  
16927     NaN  
16928     NaN  

[16929 rows x 60 columns]

In [53]:
# 계산된 성장성 데이터와 결합
df = pd.concat([kosdaq_성장성, kosdaq_1],axis=1)
df

회사명  거래소코드     회계년도      총자본증가율    유형자산증가율     유동자산증가율   비유동자산증가율  \
0      (주)CMG제약  58820  2011/12    1.975650  -5.902879   17.636684 -12.887724   
1      (주)CMG제약  58820  2012/12  109.148508  -4.130614  207.079069 -16.424273   
2      (주)CMG제약  58820  2013/12   -2.092526  16.095412  -18.890233  77.047705   
3      (주)CMG제약  58820  2014/12    3.854202  74.308191  -25.933002  68.146642   
4      (주)CMG제약  58820  2015/12   -2.643453  -5.373261   -7.169358   1.659556   
...         ...    ...      ...         ...        ...         ...        ...   
16924   흥구석유(주)  24060  2018/12   -5.888978  -0.359446  -21.190176  -2.094265   
16925   흥구석유(주)  24060  2019/12   11.242454  -7.005305   80.743108  -2.630362   
16926   흥구석유(주)  24060  2020/12    0.071718 -13.916541   22.404951  -8.205825   
16927   흥구석유(주)  24060  2021/12   -1.282832   0.629556   -4.236492   0.178466   
16928   흥구석유(주)  24060  2022/12   13.658088  -1.708923  -67.309070  51.899952   

         재고자산증가율    자기자본증가율     매출액증가율  ...     기업생애주기  통계청 한국표준산업분류 10차(대분류)  \
0      -1.437298  -2.015757  56.405286  ...        도입기                    제조업   
1       7.728764  75.174670  -8.432853  ...        도입기                    제조업   
2      56.311337  55.097971  25.512150  ...        쇠퇴기                    제조업   
3       6.365014   0.012246  11.523380  ...        성장기                    제조업   
4      11.052310  -4.481358  18.850262  ...        성장기                    제조업   
...          ...        ...        ...  ...        ...                    ...   
16924 -46.081277   0.986323   6.874490  ...  Shake-Out               도매 및 소매업   
16925  96.433378   9.311206  -9.349129  ...        성숙기               도매 및 소매업   
16926 -18.636519   3.375281 -23.777287  ...  Shake-Out               도매 및 소매업   
16927  16.168421  -0.884422  15.124865  ...        쇠퇴기               도매 및 소매업   
16928  18.339978   1.508315  10.970094  ...        성장기               도매 및 소매업   

       설립연도  상장연도  경영기간  상장기간  상장폐지일  상장폐지사유  상장폐지  상장폐지년도  
0      2001  2001    10    11    NaN     NaN     0     NaN  
1      2001  2001    11    12    NaN     NaN     0     NaN  
2      2001  2001    12    13    NaN     NaN     0     NaN  
3      2001  2001    13    14    NaN     NaN     0     NaN  
4      2001  2001    14    15    NaN     NaN     0     NaN  
...     ...   ...   ...   ...    ...     ...   ...     ...  
16924  1966  1994    52    25    NaN     NaN     0     NaN  
16925  1966  1994    53    26    NaN     NaN     0     NaN  
16926  1966  1994    54    27    NaN     NaN     0     NaN  
16927  1966  1994    55    28    NaN     NaN     0     NaN  
16928  1966  1994    56    29    NaN     NaN     0     NaN  

[16929 rows x 73 columns]

## 7. Labeling

In [54]:
# 인덱스 초기화
df.reset_index(inplace=True)

In [55]:
# 컬럼추출
df_label = df[['index','회사명', '거래소코드', '회계년도', '상장폐지일', '상장폐지']]
df_label

index       회사명  거래소코드     회계년도 상장폐지일  상장폐지
0          0  (주)CMG제약  58820  2011/12   NaN     0
1          1  (주)CMG제약  58820  2012/12   NaN     0
2          2  (주)CMG제약  58820  2013/12   NaN     0
3          3  (주)CMG제약  58820  2014/12   NaN     0
4          4  (주)CMG제약  58820  2015/12   NaN     0
...      ...       ...    ...      ...   ...   ...
16924  16924   흥구석유(주)  24060  2018/12   NaN     0
16925  16925   흥구석유(주)  24060  2019/12   NaN     0
16926  16926   흥구석유(주)  24060  2020/12   NaN     0
16927  16927   흥구석유(주)  24060  2021/12   NaN     0
16928  16928   흥구석유(주)  24060  2022/12   NaN     0

[16929 rows x 6 columns]

In [56]:
# 상장폐지된 기업 데이터만 추출
df_label= df_label[df_label['상장폐지'] == 1]
df_label= df_label.groupby('회사명')['index'].apply(list)

In [57]:
# 상장폐지된 기업의 인덱스만 추출
df_label = pd.DataFrame(df_label)
df_label

index
회사명                                                          
(주)금빛       [270, 271, 272, 273, 274, 275, 276, 277, 278, ...
(주)나노트로닉스                                          [370, 371]
(주)네오퍼플                                            [539, 540]
(주)넥스지                    [592, 593, 594, 595, 596, 597, 598]
(주)녹십자셀     [711, 712, 713, 714, 715, 716, 717, 718, 719, ...
...                                                       ...
지스마트글로벌(주)  [15874, 15875, 15876, 15877, 15878, 15879, 158...
참존글로벌(주)    [15925, 15926, 15927, 15928, 15929, 15930, 159...
태산엘시디(주)                                [16241, 16242, 16243]
한솔신텍(주)     [16658, 16659, 16660, 16661, 16662, 16663, 16664]
현진소재(주)     [16844, 16845, 16846, 16847, 16848, 16849, 168...

[154 rows x 1 columns]

In [58]:
# 상장폐지가 됐던 년도(t)에 t-1년 데이터에 1을 주는 코드
for j in df_label.index:
    for i in df.index:
        if df_label["index"][j][-1] == i:
            df.loc[i,'label'] = 1
df['label'] = df["label"].fillna(0)

In [59]:
# target 인트형 변환
df['label'] = df['label'].astype(int)

In [60]:
df

index       회사명  거래소코드     회계년도      총자본증가율    유형자산증가율     유동자산증가율  \
0          0  (주)CMG제약  58820  2011/12    1.975650  -5.902879   17.636684   
1          1  (주)CMG제약  58820  2012/12  109.148508  -4.130614  207.079069   
2          2  (주)CMG제약  58820  2013/12   -2.092526  16.095412  -18.890233   
3          3  (주)CMG제약  58820  2014/12    3.854202  74.308191  -25.933002   
4          4  (주)CMG제약  58820  2015/12   -2.643453  -5.373261   -7.169358   
...      ...       ...    ...      ...         ...        ...         ...   
16924  16924   흥구석유(주)  24060  2018/12   -5.888978  -0.359446  -21.190176   
16925  16925   흥구석유(주)  24060  2019/12   11.242454  -7.005305   80.743108   
16926  16926   흥구석유(주)  24060  2020/12    0.071718 -13.916541   22.404951   
16927  16927   흥구석유(주)  24060  2021/12   -1.282832   0.629556   -4.236492   
16928  16928   흥구석유(주)  24060  2022/12   13.658088  -1.708923  -67.309070   

        비유동자산증가율    재고자산증가율    자기자본증가율  ...  통계청 한국표준산업분류 10차(대분류)  설립연도  \
0     -12.887724  -1.437298  -2.015757  ...                    제조업  2001   
1     -16.424273   7.728764  75.174670  ...                    제조업  2001   
2      77.047705  56.311337  55.097971  ...                    제조업  2001   
3      68.146642   6.365014   0.012246  ...                    제조업  2001   
4       1.659556  11.052310  -4.481358  ...                    제조업  2001   
...          ...        ...        ...  ...                    ...   ...   
16924  -2.094265 -46.081277   0.986323  ...               도매 및 소매업  1966   
16925  -2.630362  96.433378   9.311206  ...               도매 및 소매업  1966   
16926  -8.205825 -18.636519   3.375281  ...               도매 및 소매업  1966   
16927   0.178466  16.168421  -0.884422  ...               도매 및 소매업  1966   
16928  51.899952  18.339978   1.508315  ...               도매 및 소매업  1966   

       상장연도  경영기간  상장기간  상장폐지일  상장폐지사유  상장폐지  상장폐지년도  label  
0      2001    10    11    NaN     NaN     0     NaN      0  
1      2001    11    12    NaN     NaN     0     NaN      0  
2      2001    12    13    NaN     NaN     0     NaN      0  
3      2001    13    14    NaN     NaN     0     NaN      0  
4      2001    14    15    NaN     NaN     0     NaN      0  
...     ...   ...   ...    ...     ...   ...     ...    ...  
16924  1994    52    25    NaN     NaN     0     NaN      0  
16925  1994    53    26    NaN     NaN     0     NaN      0  
16926  1994    54    27    NaN     NaN     0     NaN      0  
16927  1994    55    28    NaN     NaN     0     NaN      0  
16928  1994    56    29    NaN     NaN     0     NaN      0  

[16929 rows x 75 columns]

In [61]:
# 최종 데이터 수 
df['label'].value_counts()

0    16775
1      154
Name: label, dtype: int64

In [62]:
# 불필요 컬럼 삭제 ( 파생변수, target만들기 위해 사용했던 변수 삭제 )
df.drop(columns=['상장폐지', 'index', '상장폐지일','상장폐지사유', '상장폐지년도'],axis=1,inplace=True)

In [63]:
# 코스닥 정상기업
print("정상기업수:",df[df['label'] == 0]['회사명'].nunique())
# 코스닥 정상기업 데이터
print("정상기업 데이터수:", len(df[df['label'] == 0]))
# 코스닥 상장폐지기업
print("상장폐지기업수:",df[df['label'] == 1]['회사명'].nunique())
# 코스닥 상장폐지기업 데이터
print("상장폐지기업 데이터수:", len(df[df['label'] == 1]))

정상기업수: 1720
정상기업 데이터수: 16775
상장폐지기업수: 154
상장폐지기업 데이터수: 154


In [64]:
list = df[df['label']==1]['회사명'].tolist()

In [65]:
# 추출한 데이터를 저장할 빈 데이터프레임을 만듭니다.
result_df = pd.DataFrame()

# 리스트의 각 회사명과 데이터프레임의 회사명을 비교합니다.
for company in list:
    selected_data = df[df['회사명'] == company]
    result_df = pd.concat([result_df, selected_data])

# 결과 데이터프레임을 출력합니다.
index_list = result_df[result_df['label']== 0].index

In [66]:
# 추출한 데이터를 저장할 빈 데이터프레임을 만듭니다.
result_df = pd.DataFrame()

# 인덱스가 리스트에 포함되지 않은 데이터만 추출합니다.
for i in df.index:
    if i not in index_list:
        result_df = result_df.append(df.loc[i])

# 결과 데이터프레임을 출력합니다.
result_df

회사명  거래소코드     회계년도      총자본증가율    유형자산증가율     유동자산증가율   비유동자산증가율  \
0      (주)CMG제약  58820  2011/12    1.975650  -5.902879   17.636684 -12.887724   
1      (주)CMG제약  58820  2012/12  109.148508  -4.130614  207.079069 -16.424273   
2      (주)CMG제약  58820  2013/12   -2.092526  16.095412  -18.890233  77.047705   
3      (주)CMG제약  58820  2014/12    3.854202  74.308191  -25.933002  68.146642   
4      (주)CMG제약  58820  2015/12   -2.643453  -5.373261   -7.169358   1.659556   
...         ...    ...      ...         ...        ...         ...        ...   
16924   흥구석유(주)  24060  2018/12   -5.888978  -0.359446  -21.190176  -2.094265   
16925   흥구석유(주)  24060  2019/12   11.242454  -7.005305   80.743108  -2.630362   
16926   흥구석유(주)  24060  2020/12    0.071718 -13.916541   22.404951  -8.205825   
16927   흥구석유(주)  24060  2021/12   -1.282832   0.629556   -4.236492   0.178466   
16928   흥구석유(주)  24060  2022/12   13.658088  -1.708923  -67.309070  51.899952   

         재고자산증가율    자기자본증가율     매출액증가율  ...  투자활동으로인한현금흐름  재무활동으로인한현금흐름  \
0      -1.437298  -2.015757  56.405286  ...       -1773.0        3039.0   
1       7.728764  75.174670  -8.432853  ...      -30266.0       35361.0   
2      56.311337  55.097971  25.512150  ...       21135.0        -157.0   
3       6.365014   0.012246  11.523380  ...      -13611.0           5.0   
4      11.052310  -4.481358  18.850262  ...       -8671.0           1.0   
...          ...        ...        ...  ...           ...           ...   
16924 -46.081277   0.986323   6.874490  ...        4839.0       -7453.0   
16925  96.433378   9.311206  -9.349129  ...        -181.0       -2249.0   
16926 -18.636519   3.375281 -23.777287  ...          17.0       -2643.0   
16927  16.168421  -0.884422  15.124865  ...        2558.0       -2457.0   
16928  18.339978   1.508315  10.970094  ...      -10924.0        8112.0   

       현금및현금성자산의증가     기업생애주기  통계청 한국표준산업분류 10차(대분류)  설립연도  상장연도  경영기간  상장기간  \
0          -1295.0        도입기                    제조업  2001  2001    10    11   
1           3522.0        도입기                    제조업  2001  2001    11    12   
2          17031.0        쇠퇴기                    제조업  2001  2001    12    13   
3         -12878.0        성장기                    제조업  2001  2001    13    14   
4          -6261.0        성장기                    제조업  2001  2001    14    15   
...            ...        ...                    ...   ...   ...   ...   ...   
16924       2429.0  Shake-Out               도매 및 소매업  1966  1994    52    25   
16925      -1699.0        성숙기               도매 및 소매업  1966  1994    53    26   
16926       -905.0  Shake-Out               도매 및 소매업  1966  1994    54    27   
16927       -486.0        쇠퇴기               도매 및 소매업  1966  1994    55    28   
16928       -401.0        성장기               도매 및 소매업  1966  1994    56    29   

       label  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
16924      0  
16925      0  
16926      0  
16927      0  
16928      0  

[16299 rows x 70 columns]

In [67]:
df = result_df.reset_index(drop=True)

# 8. 상장일 이전 데이터 제거

In [68]:
# 상장일 데이터 로드 후 회사명기준으로 중복 데이터 삭제
df1 = pd.read_csv('./datasets/상장일.csv', encoding='cp949')
df1 = df1[['거래소코드', '상장일']]
df1.drop_duplicates('거래소코드', inplace=True)

In [69]:
df2 = pd.merge(df, df1, how='left', on='거래소코드')

In [70]:
df2['회계년도'] = df2['회계년도'].astype('object') + '/31'
df3 = df2[~(df2['상장일'] >= df2['회계년도'])]
df3['label'].value_counts()

0    13414
1      154
Name: label, dtype: int64

In [71]:
# 회계년도 다시 원래대로 초기화
df3['회계년도'] = pd.to_datetime(df3['회계년도'], format=('%Y/%m/%d')).dt.strftime('%Y/%m')

In [72]:
# 컬럼명 변경 ( 매출가대매출액비율 -> 매출원가대매출액비율 )
df3.rename(columns={'매출가대매출액비율' : '매출원가대매출액비율', '영업활동으로인한현금흐름' : '영업활동현금흐름',
                        '투자활동으로인한현금흐름' : '투자활동현금흐름', '재무활동으로인한현금흐름' : '재무활동현금흐름'}, inplace=True)

In [73]:
# 불필요 컬럼 삭제
df3.drop(columns=['광고선전비대매출액비율', 'R&D투자효율', '설립연도', '상장연도', '상장일'], axis=1, inplace=True)

In [74]:
df3['경영기간'] = df3['경영기간'].astype(float)
df3['상장기간'] = df3['상장기간'].astype(float)

# 9. 회사명이 2개이고 거래소코드가 1개인 기업 제거

In [75]:
# 제이콘텐트리 회사 제거
df4 = df3[df3['회사명'] != '(주)제이콘텐트리']

In [76]:
df4.reset_index(drop=True, inplace=True)

# 10. 22년도 데이터 제거

In [77]:
df5 = df4[df4['회계년도'] != '2022/12']
df5.reset_index(drop=True, inplace=True)

In [78]:
df5

회사명  거래소코드     회계년도      총자본증가율    유형자산증가율     유동자산증가율   비유동자산증가율  \
0      (주)CMG제약  58820  2011/12    1.975650  -5.902879   17.636684 -12.887724   
1      (주)CMG제약  58820  2012/12  109.148508  -4.130614  207.079069 -16.424273   
2      (주)CMG제약  58820  2013/12   -2.092526  16.095412  -18.890233  77.047705   
3      (주)CMG제약  58820  2014/12    3.854202  74.308191  -25.933002  68.146642   
4      (주)CMG제약  58820  2015/12   -2.643453  -5.373261   -7.169358   1.659556   
...         ...    ...      ...         ...        ...         ...        ...   
12091   흥구석유(주)  24060  2017/12    0.979314  -0.565031    7.340814  -0.484894   
12092   흥구석유(주)  24060  2018/12   -5.888978  -0.359446  -21.190176  -2.094265   
12093   흥구석유(주)  24060  2019/12   11.242454  -7.005305   80.743108  -2.630362   
12094   흥구석유(주)  24060  2020/12    0.071718 -13.916541   22.404951  -8.205825   
12095   흥구석유(주)  24060  2021/12   -1.282832   0.629556   -4.236492   0.178466   

         재고자산증가율    자기자본증가율     매출액증가율  ...  순운전자본회전률  영업활동현금흐름  투자활동현금흐름  \
0      -1.437298  -2.015757  56.405286  ...      0.86   -2560.0   -1773.0   
1       7.728764  75.174670  -8.432853  ...      0.42   -1573.0  -30266.0   
2      56.311337  55.097971  25.512150  ...      0.42   -3939.0   21135.0   
3       6.365014   0.012246  11.523380  ...      0.56     725.0  -13611.0   
4      11.052310  -4.481358  18.850262  ...      0.75    2410.0   -8671.0   
...          ...        ...        ...  ...       ...       ...       ...   
12091   9.408495   0.832635   8.117296  ...      7.20    1558.0      58.0   
12092 -46.081277   0.986323   6.874490  ...      8.08    5043.0    4839.0   
12093  96.433378   9.311206  -9.349129  ...      6.91     731.0    -181.0   
12094 -18.636519   3.375281 -23.777287  ...      4.00    1721.0      17.0   
12095  16.168421  -0.884422  15.124865  ...      4.44    -587.0    2558.0   

       재무활동현금흐름  현금및현금성자산의증가     기업생애주기  통계청 한국표준산업분류 10차(대분류)  경영기간  상장기간  \
0        3039.0      -1295.0        도입기                    제조업  10.0  11.0   
1       35361.0       3522.0        도입기                    제조업  11.0  12.0   
2        -157.0      17031.0        쇠퇴기                    제조업  12.0  13.0   
3           5.0     -12878.0        성장기                    제조업  13.0  14.0   
4           1.0      -6261.0        성장기                    제조업  14.0  15.0   
...         ...          ...        ...                    ...   ...   ...   
12091   -1468.0        148.0  Shake-Out               도매 및 소매업  51.0  24.0   
12092   -7453.0       2429.0  Shake-Out               도매 및 소매업  52.0  25.0   
12093   -2249.0      -1699.0        성숙기               도매 및 소매업  53.0  26.0   
12094   -2643.0       -905.0  Shake-Out               도매 및 소매업  54.0  27.0   
12095   -2457.0       -486.0        쇠퇴기               도매 및 소매업  55.0  28.0   

       label  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
12091      0  
12092      0  
12093      0  
12094      0  
12095      0  

[12096 rows x 66 columns]

In [79]:
df5.to_csv('./datasets/1.kosdaq_preprocessing완료.csv')